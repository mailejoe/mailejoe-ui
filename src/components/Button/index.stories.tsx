// Button.stories.jsx
import React from 'react';

import Button from '.';

export default {
  title: 'Components/Button',
  component: Button,
};

export const Basic = () => <Button />;
export const WithSpinner = () => <Button isWorking={true} />;