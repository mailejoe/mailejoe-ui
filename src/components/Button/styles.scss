.button {
  padding: 0 15px;
  white-space: nowrap;
  display: flex;
  align-items: center;
  justify-content: center;
  height: 28px;
  border-radius: 4px;
  background-color: #565656;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.15);
  color: #ffffff;
  font-family: 'Montersat', sans-serif;
  font-size: 12px;
  font-weight: bold;
  cursor: pointer;

  .content {
    padding: 0;
  }
}