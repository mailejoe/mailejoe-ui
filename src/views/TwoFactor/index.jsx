import React, { useEffect } from 'react';
import '../../sass/views/two-factor.scss';
import '../../sass/views/login.scss';
import axios from 'axios';

import { Formik, Form, Field } from 'formik';

import Button from '../../components/Button';
import Spinner from '../../components/Spinner';
import useTranslations from '../../translations';

const TwoFactor = () => {
  const translations = useTranslations();

  useEffect(() => {
    document.body.classList.add('auth');
    return () => {
      document.body.classList.remove('auth');
    };
  });

  const allowInput = v => {
    if (v.length !== 0) {
      if (!v.match(/^\d+$/)) {
        return false;
      }
    }

    return true;
  };

  return (
    <div className="auth">
      <div className="auth-container">
        <div className="header">
          <div className="logo">
            <img src="/images/Logo.png" style={{ height: '150px' }} />
          </div>
        </div>
        <div className="form-container">
          <Formik
            initialValues={{
              token: '',
            }}
            validate={values => {
              const errors = {};
              if (!values.token) {
                errors.token = translations.required;
              }
              return errors;
            }}
            onSubmit={async (values, { setSubmitting }) => {
              try {
                const response = await axios({
                  url: `${process.env.REACT_APP_API_URL}/api/v1/auth/mfa`,
                  method: 'POST',
                  withCredentials: true,
                  headers: {
                    'Content-Type': 'application/json',
                  },
                  data: {
                    token: values.token,
                  },
                });
                if (response.data.isValid) {
                  window.location.replace('/dashboard');
                }
                setSubmitting(false);
              } catch (err) {
                setSubmitting(false);
              }
            }}
          >
            {({ values, errors, touched, handleChange, handleBlur, handleSubmit, isSubmitting }) => (
              <Form
                className="grid-form"
                onSubmit={handleSubmit}
                autoComplete="off"
                {...(process.env.REACT_APP_IS_TEST && { 'data-testid': 'mfa:form' })}
              >
                <>
                  <>
                    <Field
                      id="token"
                      name="token"
                      type="text"
                      autoFocus={true}
                      onChange={ev => {
                        const v = ev.target.value;
                        if (!allowInput(v)) {
                          return false;
                        }
                        return handleChange(ev);
                      }}
                      onBlur={handleBlur}
                      value={values.token}
                      className={`field${touched.token && errors.token ? ' has-error' : ''}`}
                    />
                    {touched.token && errors.token && (
                      <div className="form-error" data-testid="mfa-error">
                        {errors.token}
                      </div>
                    )}
                  </>

                  <div className="actions">
                    {!isSubmitting && (
                      <Button
                        type="submit"
                        {...(process.env.REACT_APP_IS_TEST && { 'data-testid': 'mfa:submit' })}
                        color="#a44900"
                      >
                        {translations.validate}
                      </Button>
                    )}
                    {isSubmitting && (
                      <div className="loading">
                        <Spinner />
                      </div>
                    )}
                  </div>

                  <div className="divider" />
                </>
              </Form>
            )}
          </Formik>
        </div>
      </div>
      <div className="auth-bg"></div>
    </div>
  );
};

export default TwoFactor;
