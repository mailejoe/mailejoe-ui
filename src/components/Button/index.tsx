import React, { ReactChildren, ReactNode } from 'react';
import './styles.scss';

import Spinner from '../Spinner';

interface ButtonProps {
  className?: string;
  children?: ReactChildren | ReactNode;
  iconSize?: number;
  color?: string;
  type?: 'button' | 'submit' | 'reset';
  disabled?: boolean;
  loading?: boolean;
  isWorking?: boolean,
  onClick?: () => void;
  testId?: string;
}

const defaultProps = {
  className: undefined,
  children: undefined,
  iconSize: 24,
  color: 'white',
  disabled: false,
  loading: false,
  isWorking: false,
  onClick: () => {},
};

export const Button = (props: ButtonProps = defaultProps) => {
  const { className, loading, type, disabled, iconSize, onClick, color, children, testId } = props;
  
  const handleClick = () => {
    if (!disabled && !loading) {
      if (onClick) onClick();
    }
  };

  return (
    <button
      {...process.env.REACT_APP_IS_TEST && { 'data-testid': testId }}
      className={`button ${className ? className : ''}`}
      onClick={handleClick}
      disabled={disabled || loading}
      {...type && { type }}
    >
      {loading && <Spinner size={iconSize} color={color} />}
      {children && <div className="content">{children}</div>}
    </button>
  );
}

export default Button;