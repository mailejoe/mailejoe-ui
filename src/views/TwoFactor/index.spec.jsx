import { configure, mount } from 'enzyme';
import React from 'react';
import TwoFactor from '.';
import { Form, Field } from 'formik';
import Adapter from 'enzyme-adapter-react-16';
import Button from '../../components/Button';
import axios from 'axios';

jest.mock('axios');

configure({ adapter: new Adapter() });

let wrapper;
describe('<TwoFactor />', () => {
  beforeEach(() => {
    wrapper = mount(<TwoFactor />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should render snapshot', () => {
    expect(TwoFactor).toMatchSnapshot();
  });

  it('should do a basic render', () => {
    expect(wrapper.find('.auth-container')).toHaveLength(1);
    expect(wrapper.find('.header')).toHaveLength(1);
    expect(wrapper.find('.auth-bg')).toHaveLength(1);
    expect(wrapper.find('.logo')).toHaveLength(1);
    expect(wrapper.find('.form-container')).toHaveLength(1);
    expect(wrapper.find(Form)).toHaveLength(1);
    expect(wrapper.find('img')).toHaveLength(1);

    expect(document.body.className).toBe('auth');
  });

  describe('TwoFactor Form', () => {
    it('should render the two factor form', () => {
      const form = wrapper.find(Form);
      expect(form).toHaveLength(1);
      expect(form.props().autoComplete).toBe('off');
      expect(form.props()['data-testid']).toBe('mfa:form');

      const formInputs = wrapper.find(Field);
      expect(formInputs).toHaveLength(1);

      const token = formInputs.at(0);
      expect(token.props().name).toBe('token');
      expect(token.props().type).toBe('text');
      expect(token.props().className).toBe('field');

      expect(wrapper.find('.form-error')).toHaveLength(0);

      const validateBtn = wrapper.find(Button);
      expect(validateBtn).toHaveLength(1);
      expect(validateBtn.props().color).toBe('#a44900');
      expect(validateBtn.props()['data-testid']).toBe('mfa:submit');
      expect(validateBtn.text()).toBe('Validate');

      expect(wrapper.find('.divider')).toHaveLength(1);
    });

    it('should handle an empty token', () => {
      const validateBtn = wrapper.find(Button);

      validateBtn.find('.button').simulate('click');

      setTimeout(() => {
        expect(wrapper.find('.form-error')).toHaveLength(1);
      }, 100);
    });

    it('should handle a change after validation', () => {
      const validateBtn = wrapper.find(Button);

      validateBtn.find('.button').simulate('click');

      setTimeout(() => {
        const formErrors = wrapper.find('.form-error');
        expect(formErrors).toHaveLength(1);
        expect(formErrors.at(0).text).toEqual('Required');

        const formInputs = wrapper.find(Field);
        expect(formInputs).toHaveLength(1);

        const token = formInputs.at(0);
        token.simulate('change', { target: { value: 'Hello' } });

        setTimeout(() => {
          expect(wrapper.find('.form-error')).toHaveLength(0);
        }, 100);
      }, 100);
    });
  });
});
