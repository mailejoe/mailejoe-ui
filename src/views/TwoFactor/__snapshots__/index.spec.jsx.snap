// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TwoFactor /> should render snapshot 1`] = `[Function]`;
